### PR TITLE
Update to CUTLASS v3.4.0

### PR DESCRIPTION
**Pull request by cpuhrsch**
Update from 3.2.2 to 3.4.0

Some highlights useful for the sparse / quantization / nested workstream.

https://github.com/NVIDIA/cutlass/releases/tag/v3.4.0
- Improved [Mixed-input Hopper GEMMs](https://github.com/NVIDIA/cutlass/blob/main/examples/55_hopper_mixed_dtype_gemm) supporting {16-bit, 8-bit} x {8-bit, 4-bit} input types with fast numerical converters and group scaling factors tuned for optimal performance on Hopper H100.
- Beta release of [Pointer-Array Batched GEMMs](https://github.com/NVIDIA/cutlass/blob/main/examples/56_hopper_ptr_array_batched_gemm) utilizing TMA and Hopper H100 tensor cores now available. (Requires CUDA 12.3 or above)
- Beta release of [Group-GEMM](https://github.com/NVIDIA/cutlass/blob/main/examples/57_hopper_grouped_gemm) - commonly used in optimization of Mixture-Of-Expert models, is now available on Hopper GPUs taking advantage of TMA and Hopper H100 tensor cores. (Requires CUDA 12.3 or above)
- [Ampere Sparse GEMM](https://github.com/NVIDIA/cutlass/blob/main/examples/15_ampere_sparse_tensorop_gemm/ampere_sparse_tensorop_gemm_with_visitor.cu) supports Epilogue Visitor Tree (EVT) now.

https://github.com/NVIDIA/cutlass/releases/tag/v3.3.0
- New https://github.com/NVIDIA/cutlass/pull/1084 with support for canonical layouts (TN). The implementation supports upcast on operandB {fp16, bf16} x {s8, u8} and upcast on operandA {s8, u8} x {fp16, bf16}. They also include fast numeric conversion recipes and warp level shuffles to achieve optimal performance.

PRs blocked on this update
- https://github.com/pytorch/pytorch/pull/117124

### CODE REVIEW

**Comment by cpuhrsch** [comment]
cc @alexsamardzic

**Comment by cpuhrsch** [comment]
```
2024-01-18T20:34:15.5247881Z 2 errors detected in the compilation of "/var/lib/jenkins/workspace/aten/src/ATen/native/transformers/cuda/attention.cu".
2024-01-18T20:34:15.5250711Z make[2]: *** [caffe2/CMakeFiles/torch_cuda.dir/build.make:4286: caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/transformers/cuda/attention.cu.o] Error 1
2024-01-18T20:34:15.6769591Z /var/lib/jenkins/workspace/aten/src/ATen/../../../third_party/cutlass/include/cute/atom/mma_atom.hpp(216): error: static assertion failed with "TiledMMA requires independent permutations of MNK."
2024-01-18T20:34:15.6770960Z           detected during:
2024-01-18T20:34:15.6773031Z             instantiation of class "cute::TiledMMA<MMA_Atom, AtomLayoutMNK, PermutationMNK> [with MMA_Atom=cute::MMA_Atom<cute::SM75_16x8x8_F32F16F16F32_TN>, AtomLayoutMNK=cute::Layout<cute::tuple<cute::C<4>, cute::C<2>, cute::_1>, cute::tuple<cute::_1, cute::_4, cute::C<0>>>, PermutationMNK=cute::Layout<cute::tuple<cute::_1, cute::_2, cute::_2>, cute::tuple<cute::C<0>, cute::C<1>, cute::C<2>>>]" 
2024-01-18T20:34:15.6776724Z /var/lib/jenkins/workspace/aten/src/ATen/native/transformers/cuda/flash_attn/flash_bwd_kernel.h(278): here
2024-01-18T20:34:15.6779174Z             instantiation of "void pytorch_flash::convert_dQ<Kernel_traits,Params>(const Params &) [with Kernel_traits=pytorch_flash::Flash_bwd_kernel_traits<160, 64, 64, 8, 4, 4, 4, false, false, cutlass::half_t, pytorch_flash::Flash_kernel_traits<160, 64, 64, 8, cutlass::half_t>>, Params=pytorch_flash::Flash_bwd_params]" 
2024-01-18T20:34:15.6781701Z /var/lib/jenkins/workspace/aten/src/ATen/native/transformers/cuda/flash_attn/flash_bwd_launch_template.h(45): here
2024-01-18T20:34:15.6784026Z             instantiation of "void pytorch_flash::flash_bwd_convert_dq_kernel<Kernel_traits>(pytorch_flash::Flash_bwd_params) [with Kernel_traits=pytorch_flash::Flash_bwd_kernel_traits<160, 64, 64, 8, 4, 4, 4, false, false, cutlass::half_t, pytorch_flash::Flash_kernel_traits<160, 64, 64, 8, cutlass::half_t>>]" 
2024-01-18T20:34:15.6786363Z /var/lib/jenkins/workspace/aten/src/ATen/native/transformers/cuda/flash_attn/flash_bwd_launch_template.h(89): here
2024-01-18T20:34:15.6789084Z             instantiation of "void pytorch_flash::run_flash_bwd_seqk_parallel<Kernel_traits,Is_dropout>(pytorch_flash::Flash_bwd_params &, cudaStream_t, __nv_bool) [with Kernel_traits=pytorch_flash::Flash_bwd_kernel_traits<160, 64, 64, 8, 4, 4, 4, false, false, cutlass::half_t, pytorch_flash::Flash_kernel_traits<160, 64, 64, 8, cutlass::half_t>>, Is_dropout=true]" 
2024-01-18T20:34:15.6791547Z /var/lib/jenkins/workspace/aten/src/ATen/native/transformers/cuda/flash_attn/flash_bwd_launch_template.h(141): here
2024-01-18T20:34:15.6793985Z             instantiation of "void pytorch_flash::run_flash_bwd<Kernel_traits,Is_dropout>(pytorch_flash::Flash_bwd_params &, cudaStream_t, __nv_bool) [with Kernel_traits=pytorch_flash::Flash_bwd_kernel_traits<160, 64, 64, 8, 4, 4, 4, false, false, cutlass::half_t, pytorch_flash::Flash_kernel_traits<160, 64, 64, 8, cutlass::half_t>>, Is_dropout=true]" 
2024-01-18T20:34:15.6796492Z /var/lib/jenkins/workspace/aten/src/ATen/native/transformers/cuda/flash_attn/flash_bwd_launch_template.h(294): here
2024-01-18T20:34:15.6798013Z             instantiation of "void pytorch_flash::run_mha_bwd_hdim160<T>(pytorch_flash::Flash_bwd_params &, cudaStream_t, __nv_bool) [with T=cutlass::half_t]" 
2024-01-18T20:34:15.6799588Z /var/lib/jenkins/workspace/aten/src/ATen/native/transformers/cuda/flash_attn/kernels/flash_bwd_hdim160_fp16_sm80.cu(12): here
```

https://hud.pytorch.org/pr/117760

Looks like there are some first order incompatibilities with flash attention.

**Comment by cpuhrsch** [comment]
I see that FlashAttention was updated against CUTLASS 3.3 but, from what I can tell, not yet against 3.4: https://github.com/Dao-AILab/flash-attention/commit/8448c02889542f720880b3afcd0aedb0e9ab164e

**Review by Skylion007 — APPROVED** [review]

**Comment by kadeng** [comment]
See also a PR where I try to upgrade to Cutlass 3.3.0: https://github.com/pytorch/pytorch/pull/118629
I would also prefer going directly to 3.4.0, but maybe taking that intermediate step makes sense?

**Comment by fxmarty** [comment]
@cpuhrsch Will mixed fp16 @ int8 be exposed through pytorch?

**Comment by alexsamardzic** [comment]
> @cpuhrsch Will mixed fp16 @ int8 be exposed through pytorch?

This is already supported in PyTorch through `_mixed_dtypes_linear()` operator - check `test_mixed_dtypes_linear` unit test in `test/test_matmul_cuda.py`.  Admittedly, the functionality is not based directly on CUTLASS, but there is a [PR](https://github.com/pytorch/pytorch/pull/110600) that actually uses CUTLASS int8 support (that is better in some aspects, in particular in that it doesn't require reordering of the integer tensor up-front) for mixed data-types GEMM for the implementation.  However, CUTLASS doesn't yet supports int4/uint4, so there is also [PR](https://github.com/NVIDIA/cutlass/pull/1190) to add support for these in CUTLASS.  Thus, CUTLASS PR is to be merged first, and then PyTorch PR will be updated for 4-bit integer support, and then merged too; but in the meantime you should be able to use existing implementation of this functionality.

**Comment by cpuhrsch** [comment]
@fxmarty - The public API for this is exposed via torch.compile, but I understand that we'd probably also want to design an API extension (perhaps similar to [out_dtype](https://github.com/pytorch/pytorch/blob/4e93b00b692118b8531f3807ec95eb4c538ea419/torch/_higher_order_ops/out_dtype.py#L107)) for torch.(b)m(at)m(ul).

However, for CUTLASS kernels to have best performance we really need to tune the kernel hyperparameters across a larger grid. torch.compile's autotuner can help with this and @alexsamardzic is looking into extensions to support this.
